server/src/main/java/org/apache/iotdb/db/engine/migration/MigrationManager.java

```diff
@@ -0,0 +1,478 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.db.engine.migration;
+
+import org.apache.iotdb.db.concurrent.IoTDBThreadPoolFactory;
+import org.apache.iotdb.db.conf.IoTDBConfig;
+import org.apache.iotdb.db.conf.IoTDBDescriptor;
+import org.apache.iotdb.db.engine.StorageEngine;
+import org.apache.iotdb.db.engine.migration.MigrationTaskWriter.MigrationLog;
+import org.apache.iotdb.db.metadata.path.PartialPath;
+import org.apache.iotdb.db.qp.utils.DatetimeUtils;
+import org.apache.iotdb.tsfile.fileSystem.FSFactoryProducer;
+import org.apache.iotdb.tsfile.utils.FilePathUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * MigrationManager keep tracks of all Migration Tasks, creates the threads to check/run the
+ * MigrationTask.
+ */
+public class MigrationManager {
+  private static final Logger logger = LoggerFactory.getLogger(MigrationManager.class);
+  protected static IoTDBConfig config = IoTDBDescriptor.getInstance().getConfig();
+  private MigrationTaskWriter logWriter;
+
+  // taskId -> MigrationTask
+  private ConcurrentHashMap<Long, MigrationTask> migrationTasks = new ConcurrentHashMap<>();
+  // the current largest MigrationTask id + 1, used to create new tasks
+  private long currentTaskId = 0;
+  // single thread to iterate through migrationTasks and check start
+  private ScheduledExecutorService migrationCheckThread;
+  // multiple threads to run the tasks
+  private ExecutorService migrationTaskThreadPool;
+
+  private boolean initialized = false;
+  private static final long MIGRATE_CHECK_INTERVAL = 60 * 1000L;
+  private static final String LOG_FILE_NAME =
+      Paths.get(FilePathUtils.regularizePath(config.getSystemDir()), "migration", "log.bin")
+          .toString();
+
+  protected MigrationManager() {
+    init();
+  }
+
+  // singleton
+  private static class MigrationManagerHolder {
+    private MigrationManagerHolder() {}
+
+    private static final MigrationManager INSTANCE = new MigrationManager();
+  }
+
+  public static MigrationManager getInstance() {
+    return MigrationManagerHolder.INSTANCE;
+  }
+
+  public synchronized void init() {
+    if (initialized) {
+      return;
+    }
+
+    try {
+      logWriter = new MigrationTaskWriter(LOG_FILE_NAME);
```

Comment:
Please close this when the current system exits.

Reply:
Fixed.